daemon/algod/api/server/router.go

```diff
@@ -107,8 +142,7 @@ func NewRouter(logger logging.Logger, node *node.AlgorandFullNode, shutdown <-ch
 	e.Pre(middleware.RemoveTrailingSlash())
 
 	e.Use(echo.WrapMiddleware(middlewares.Logger(logger)))
-	// TODO: New auth middleware - https://github.com/algorand/go-algorand/issues/863
-	e.Use(middlewares.Auth(logger, apiToken))
+	e.Use(middlewares.Auth(logger, makeAuthRoutes(apiToken, adminAPIToken, node.Config().EnableProfiler)))
```

Comment:
Rather than attaching the Auth middleware globally, it's possible to attach it to specific sub-routes. So for example we could skip adding it to the common routes.

I think that you're implementation is fine, but this is another option.

Reply:
The collector pattern is very clever, but I think we can avoid it. I could also update our `oapi-codgen` fork so that we could pass middleware to the handlers which could simplify some of the code:
```
generated.RegisterHandlers(router, handler, authMiddleware(generalKey))
private.RegisterHandlers(router, handler, authMiddleware(privatekey))
```

We could already do this with the pprof endpoints:
```
e.GET("/debug/pprof/*", echo.WrapHandler(http.DefaultServeMux), authMiddleware(private))
e.GET("/urlAuth/:token/debug/pprof/*", echo.WrapHandler(http.DefaultServeMux), authMiddleware(private)
```

Reply:
Yes, I was thinking about refactoring a better code to start with, but figured out that we could push that for later. As long as it's behaves correctly, we can change the underlaying implementation later on.